### PR TITLE
Recipe App 

### PR DESCRIPTION
**Pull request by 0sugo**
In this Project @skabeo and @0sugo  did the following :

1. Set up the project using `rails new` command
2. Set up User authentication using the `Devise` gem
3. Listed food belonging to the logged in user
4. Allowed logged in user to add a food
5. Displayed recipes .When the recipe title is clicked the recipe details is displayed if the logged in user is the creator or the recipe is set to public.
6. Configured authorisation using `cancancan` to ensure only the owner of a recipe can perform changes to it
7. Displayed public recipes with their information (The owner,Total recipe foods and the total price)
8. In the recipe details,Owners can add ingredients,modify or delete ingredients
9. Ensured on the recipe details page,Users can generate a shopping list for that specific recipe.
10. Created a general shopping list to display the total quantity,total price and name of ingredient for a specific recipe
11. Fixed N+1 queries
12. created a home page that gave access to the following pages:
    - view all foods
    - view recipes
    -  view public recipes

13. Wrote unit tests for models which included:
    - User
    - Recipe
    - Recipe food
    - Food

18. Wrote integration tests for:
    - Food index view page
    - Recipe index view page
    - Recipe show view page
    - Shopping list view page
    - Public recipe index view page

### CODE REVIEW

File: spec/models/food_spec.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
+require 'rails_helper'
+require 'shoulda/matchers'
+
+RSpec.describe Food, type: :model do
+  describe 'Food model' do
+    describe 'associations' do
+      it { should belong_to(:user) }
+      it { should have_many(:recipe_foods).dependent(:destroy) }
```
Comment (zieeco) [inline spec/models/food_spec.rb:8]: - Hi, these lines are making your test fails in three cases. Please, verify to fix it.


image

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'associations' do
+    it { should have_many(:foods) }
+    it { should have_many(:recipes) }
```
Comment (zieeco) [inline spec/models/user_spec.rb:6]: - The same issue is applicable here and in line 16. Please, verify to fix it.

image

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @Team,



Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- No linter errors.✔️
- Nice design. ✔️
- Descriptive PR. ✔️
- Console app works correctly :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._




## Optional suggestions

- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or commthat the code review was not fairents in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb (hunk 2 above)
Comment (0sugo) [reply spec/models/user_spec.rb]: Thanks @zieeco .Resolved

File: spec/models/food_spec.rb (hunk 1 above)
Comment (0sugo) [reply spec/models/food_spec.rb]: Thanks @zieeco .Resolved

**Review by christianbiring1 — APPROVED** [review]
Hi @0sugo , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
# APPROVED :tada: 
## Highlights
 - [x] Great work on implementing the changes as suggested by the previous reviewer :+1: 
 - [x] All tests cases are passing :heavy_check_mark: :100: 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**